shiftcommerce/flex-ruby-gem
# Issue: Add TaxCode feature

### Overview

Related GitHub issue: #127

PR adds the TaxCode feature. 

### QA

1.) Add branch to local F/E, eg. 

```ruby
gem "flex_commerce_api", git: "https://github.com/shiftcommerce/flex-ruby-gem.git", branch: "issue/127_tax_code_model"
``` 

*Note: Ensure your API key has permissions  to read TaxCodes.


2.) Query the TaxCode API via the Flex gem

```ruby
$ rails c

### Create TaxCode

$ tax_code_attributes = {
  code: rand(32**8).to_s(32), 
  starts_at: Time.now, 
  ends_at: 2.days.from_now,
  country: "GB",
  rate: 0.1
}

$ new_record = FlexCommerce::TaxCode.create(tax_code_attributes)

$ new_record.persisted?
=> true


### Fetch the TaxCode collection and check the fetched record is present

$ FlexCommerce::TaxCode.all



### Fetch the created TaxCode

$ fetched_record = FlexCommerce::TaxCode.find(THE_ID)



### Update the TaxCode with invalid attributes

$ invalid_attrs = { code: "" }

$ fetched_record.update_attributes(invalid_attrs)
=> false

$ fetched_record.errors
=> #<ActiveModel::Errors:0x007ffc9e1da400
 @base=
  #<FlexCommerce::TaxCode:@attributes={"id"=>"13", "type"=>"tax_codes", "updated_at"=>"2018-04-17T08:36:24Z", "created_at"=>"2018-04-17T08:36:24Z", "code"=>"", "country"=>"GB", "rate"=>0.1, "starts_at"=>"2018-04-17T08:36:12Z", "ends_at"=>"2018-04-19T08:36:12Z"}>,
 @details={:code=>[{:error=>"can't be blank"}]},
 @messages={:code=>["can't be blank"]}>




### Update the TaxCode with valid attributes

$ valid_attrs = { code: "new code" }

$ fetched_record.update_attributes(valid_attrs)
=> true



### Fetch the updated TaxCode

$ updated_record = FlexCommerce::TaxCode.find(fetched_record.id)

$ updated_record.code
=> "new code"



### Delete the TaxCode

$ updated_record_id = updated_record.id

$ updated_record.destroy
=> true



### Check the TaxCode has been deleted

$ FlexCommerce::TaxCode.find(updated_record_id)
=> FlexCommerceApi::Error::NotFound: Couldn't find resource at:...



### Fetch the TaxCode collection and check the record is not present

$ FlexCommerce::TaxCode.all
```
